src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpRetry.cs

```diff
@@ -0,0 +1,248 @@
+        {
+            return false;
+        }
+
+        var delayDuration = throttleDelay.HasValue
+            ? throttleDelay.Value
+            : TimeSpan.FromMilliseconds(GetRandomNumber(0, nextRetryDelayMilliseconds));
+
+        if (deadline.HasValue && IsDeadlineExceeded(deadline + delayDuration))
+        {
+            return false;
+        }
+
+        if (throttleDelay.HasValue)
+        {
+            try
+            {
+                nextRetryDelayMilliseconds = Convert.ToInt32(throttleDelay.Value.TotalMilliseconds);
+            }
+            catch (OverflowException)
+            {
+                nextRetryDelayMilliseconds = MaxBackoffMilliseconds;
+            }
+        }
+
+        nextRetryDelayMilliseconds = CalculateNextRetryDelay(nextRetryDelayMilliseconds);
+        retryResult = new RetryResult(throttleDelay.HasValue, delayDuration, nextRetryDelayMilliseconds);
+        return true;
+    }
+
+    private static bool IsDeadlineExceeded(DateTime? deadline)
+    {
+        // This implementation is internal, and it is guaranteed that deadline is UTC.
+        return deadline.HasValue && deadline <= DateTime.UtcNow;
+    }
+
+    private static int CalculateNextRetryDelay(int nextRetryDelayMilliseconds)
+    {
+        var nextMilliseconds = nextRetryDelayMilliseconds * BackoffMultiplier;
+        nextMilliseconds = Math.Min(nextMilliseconds, MaxBackoffMilliseconds);
+        return Convert.ToInt32(nextMilliseconds);
+    }
+
+    private static TimeSpan? TryGetGrpcRetryDelay(StatusCode statusCode, Metadata trailers)
+    {
+        Debug.Assert(trailers != null, "trailers was null");
+
+        if (statusCode != StatusCode.ResourceExhausted && statusCode != StatusCode.Unavailable)
+        {
+            return null;
+        }
+
+        var statusDetails = trailers.Get(GrpcStatusDetailsHeader);
+        if (statusDetails != null && statusDetails.IsBinary)
+        {
+            var status = Status.Parser.ParseFrom(statusDetails.ValueBytes);
+            foreach (var item in status.Details)
+            {
+                var success = item.TryUnpack<RetryInfo>(out var retryInfo);
+                if (success)
+                {
+                    return retryInfo.RetryDelay.ToTimeSpan();
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private static TimeSpan? TryGetHttpRetryDelay(HttpStatusCode statusCode, HttpResponseHeaders headers)
+    {
+        Debug.Assert(headers != null, "headers was null");
+
+#if NETSTANDARD2_1_OR_GREATER || NET6_0_OR_GREATER
+        return statusCode == HttpStatusCode.TooManyRequests || statusCode == HttpStatusCode.ServiceUnavailable
+#else
+        return statusCode == (HttpStatusCode)429 || statusCode == HttpStatusCode.ServiceUnavailable
+#endif
+            ? headers.RetryAfter?.Delta
+            : null;
+    }
+
+    private static bool IsGrpcStatusCodeRetryable(StatusCode statusCode, bool hasRetryDelay)
+    {
+        switch (statusCode)
```

Comment:
@alanwest Is it possible to do this custom retry for `Unavailable` and `ResourceExhausted` only? And rest with https://learn.microsoft.com/en-us/aspnet/core/grpc/retries?view=aspnetcore-7.0#configure-a-grpc-retry-policy

Reply:
Probably not, especially if we introduce a max number of retries. There's no way to know how many attempts the client library attempted vs. special handling for unavailable/resource exhausted.

Though, even if it were possible, I'd be hard pressed to say the complexity would be buying us anything. I think it's pretty nice that gRPC and HTTP uses the exact same retry logic.